pkg/storage/cacher.go

```diff
@@ -402,10 +403,20 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 			glog.Errorf("invalid object for filter: %v", obj)
 			return false
 		}
-		if !strings.HasPrefix(objKey, key) {
-			return false
+
+		segments := strings.Split(objKey, "/")
+		currPath := ""
+
+		match := false
+		for _, segment := range segments {
+			currPath = path.Join(currPath, segment)
+			if currPath == key {
```

Comment:
Well, this is not true for exact match.
For exact match, i.e. filter("/abc", "/abc"), it will be ignored in WatchList(). See [etcd_helper](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_helper.go#L250) and [etcd3/watcher](https://github.com/kubernetes/kubernetes/blob/da7e9783e8f4793d4d0dd54dcc68f86bb23df3fb/pkg/storage/etcd3/watcher.go#L80-L82).